src/bot/handlers.py

```diff
@@ -47,42 +48,74 @@ async def start(update: Update, context: CallbackContext) -> None:
+                web_app=WebAppInfo(url=f"{settings.registration_template_url}"),
             )
         ),
     )
 
 
-async def web_app_data(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
-    """Получение данных из формы регистрации. Создание объекта User и Request."""
+async def update_user_data(
+    update: Update,
+    context: CallbackContext,
+) -> None:
+    """Инициализация формы для обновления регистрационных данных пользователя."""
+    user = context.user_data.get('user')
+    web_hook = UserWebhookTelegram.from_orm(user)
+    query = urllib.parse.urlencode(web_hook.dict())
+    await update.message.reply_text(
+        "Нажмите на кнопку ниже, чтобы перейти на форму регистрации.",
+        reply_markup=ReplyKeyboardMarkup.from_button(
+            KeyboardButton(
+                text="Подать заявку на участие в смене",
+                web_app=WebAppInfo(url=f"{settings.registration_template_url}?{query}"),
+            )
+        ),
+    )
+
+
+async def web_app_data(update: Update, context: CallbackContext) -> None:
+    """Получение данных из формы регистрации. Создание (обновление) объекта User и Request."""
     user_data = json.loads(update.effective_message.web_app_data.data)
     try:
         user_scheme = UserCreateRequest(**user_data)
-        user_scheme.telegram_id = update.effective_user.id
-        session = get_session()
-        registration_service = await get_user_service_callback(session)
-        await registration_service.register_user(user_scheme)
-        await update.message.reply_text(
-            text="Процесс регистрации занимает некоторое время - вам придет уведомление",
-            reply_markup=ReplyKeyboardRemove(),
-        )
-    except (ValidationError, ValueError) as e:
-        if isinstance(e, ValidationError):
-            e = "\n".join(tuple(error.get("msg", "Проверьте правильность заполнения данных.") for error in e.errors()))
+    except ValidationError as e:
+        e = "\n".join(tuple(error.get("msg", "Проверьте правильность заполнения данных.") for error in e.errors()))
         await update.message.reply_text(f"Ошибка при заполнении данных:\n{e}")
+        return
+    user_scheme.telegram_id = update.effective_user.id
+    session = get_session()
+    registration_service = await get_user_service_callback(session)
+    text = "Процесс регистрации занимает некоторое время - вам придет уведомление."
+    if context.user_data.get('user'):
+        text = (
+            " Обновленные данные приняты!\nПроцесс обработки заявок занимает некоторое время - вам придет уведомление."
+        )
+    await update.message.reply_text(
+        text=text,
+        reply_markup=ReplyKeyboardRemove(),
+    )
+    try:
+        await registration_service.register_user(user_scheme)
     except RegistrationException as e:
         await update.message.reply_text(text=e.detail, reply_markup=ReplyKeyboardRemove())

```

Comment:
Здесь отправку сообщения нужно перенести в самый низ, т.е. когда уже создан новый пользователь, обновлены его данные или найдена ошибка. 
Только нужно пройти по всем методам, которые вызываются в registration_service.register_user и проверить, что все вызываемые исключения наследуются от ApplicationException. Если нет, то надо исправить. Тогда всё будет работать правильно.
```suggestion
    try:
        await registration_service.register_user(user_scheme)
    except ApplicationException as e:
        text=e.detail
    await update.message.reply_text(
        text=text,
        reply_markup=ReplyKeyboardRemove(),
    )

```

Reply:
Судя по методам, то вызывается только два типа исключения: 
- RegistrationException, наследованное от starlette.HTTPException при добавлении пользователя в смену;
- ValueError при валидации данных пользователя.
В общем, вернул ValueError и RegistrationException